### DISCUSSION TITLE
Support Scriptlets

### DISCUSSION THREAD

**Discussion by yvonneekanim** [specify/specify7#5531] [Feature Requests]
The following scriptlets do not work in Sp7:
•	aggregatePrepCounts
•	aggregatePreps
•	aggregatePreps
•	buildLocalityString
•	buildNameString
•	degrees
•	format
•	formatDate
•	formatDate
•	formatDetermination
•	formatFieldNo
•	formatTaxonWithAuthors
•	getAggregated
•	getByLoanAgentRole
•	getByRole
•	getCollectors
•	getCollectorsByCOIdNumWithAggregator
•	getCollectorsWithAggregator
•	getCurrentDate
•	getCurrentDeterminationFullName
•	getFirstCollector
•	getSecondaryCollectors
•	getTaxonNameWithAuthors
•	getTypeStatus
•	getTypeTaxon
•	loanCategory
•	locality
•	localityBD
•	sumPrepCounts
•	toProperCase

**Comment by FedorSteeman** [comment]
A certain Alexi Sommerville promised me a workaround to this issue half a year ago (see attached mail). 

Is there any progress in this issue? We need "getByLoanAgentRole" to work in Specify7. 

[Re Specify 7 Report runner.msg.pdf](https://github.com/specify/specify7/files/7561478/Re.Specify.7.Report.runner.msg.pdf)

**Comment by maxpatiiuk** [comment]
From @grantfitzsimmons:
> Some scriptlets (namely, GetByLoanAgentRole) don't function in SP7 and prevent us from being able to generate loan reports that include names of people involved with shipping, prepping, and receiving the loans. (see attached loan invoice header generated in SP6). My request is that these be made to work in SP7.
> 
> ![image](https://user-images.githubusercontent.com/37256050/168164764-126890e2-0a29-431f-8045-daf8a512299c.png)
> 
> **Requested By:** [Plarson on the Specify Forum](https://discourse.specifysoftware.org/t/scriptlets-for-reports-in-sp7/496)

**Comment by grantfitzsimmons** [comment]
Add GetByLoanAgentRole to the list of scriplets

**Comment by grantfitzsimmons** [comment]
getByLoanAgentRole has been requested by RBGE as well.

**Comment by grantfitzsimmons** [comment]
Ben Ritchie from RBGE has mentioned again the importance of supporting these scriplets

**Comment by grantfitzsimmons** [comment]
Cornell Birds is waiting for `getByLoanAgentRole` before they can use Specify 7.

**Comment by FedorSteeman** [comment]
@grantfitzsimmons  I made a workaround a year or so ago that I might be willing to dig up and share.

**Comment by grantfitzsimmons** [comment]
@FedorSteeman It would be super helpful if you could share that!

**Comment by FedorSteeman** [comment]
@grantfitzsimmons 

OK I uploaded some of the relevant app resources to our github, since we need them there anyway to control version history. 

https://github.com/NHMDenmark/DanSpecify/tree/master/App%20Resources/MSJN/Paleontology 

Please take a look at e.g. the "[MSJN Loan Receipt.xml](https://github.com/NHMDenmark/DanSpecify/blob/master/App%20Resources/MSJN/Paleontology/MSJN%20Loan%20Receipt.xml)" file. 

At line 40, you will see it queries the aggregated field of all loanAgents: 

`<field name="52,53-loanagents.loanagent.loanAgents" class="java.lang.String"/>`

At line 344-347 you will see the field expression that I used to tease out the agent from the "loanAgents" field. Depending on role, it will pick out the correct agent. It also inserts line breaks for the respective elements of the agent aggregated string, by replacing pipes and removes the role name that was added during aggregation. This is repeated for other roles at their respective positions. 

```
($F{52,53-loanagents.loanagent.loanAgents}.split(";")[1].contains("authorizer") ? 
($F{52,53-loanagents.loanagent.loanAgents}.split(";")[1].replace("|", "<br/>"))
.replace("- authorizer", "")
: "- No authorizer defined! - ")
```

I know it's a bit technical, but I hope you can figure this out and know how to add it to JasperReports (or directly to the XML). 

Of course, it's important that all roles are filled out, otherwise an error will be thrown.

**Comment by grantfitzsimmons** [comment]
@FedorSteeman I appreciate your solution! I think that's a really great way to approach it and something I will share with the team today.

**Comment by specifysoftware** [comment]
This issue has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/creating-reports-labels-in-specify-7-jaspersoft-studio/628/1

**Comment by specifysoftware** [comment]
This issue has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/jaspersoft-studio-cant-resolve-edu-ku-brc-specify-config-scriptlet/1281/2

**Comment by grantfitzsimmons** [comment]
> > Agents: We need to be able to pull out details from the Authorizer, Collector, Donor, etc. Currently, we have it pulling all Accession Agents aggregated and then these could be parsed. But how do we then link those (e.g., Bauer, Jennifer) to the Agent Address and other details from the core Agent Table?

>  
> At this time, these must be aggregated together since it would require a scriptlet that serves this purpose: `getByLoanAgentRole`

**Reported by:** University of Michigan

**Comment by FedorSteeman** [comment]
Since the server migration & Specify7 update for the Copenhagen site, we've been encouraging users to completely switch away from Specify6. However, the lack of support of Scriptlets, especially getByLoanAgentRole, is barring that. Is there any news on having these implemented, or shall I try to (painstakingly) apply the solution I came up with for the other institution (mentioned above)?

**Comment by grantfitzsimmons** [comment]
@FedorSteeman Unfortunately, we have not made significant progress in reimplementing or supporting all scriptlets in Specify 7. Specify 6 often performed additional queries when scriptlets were called to retrieve data that was not returned in the original associated query, and we have been primarily against this solution. 

We've used a solution similar to the scriptlet you provided as our current solution to avoid using the previous approach, though we intend to eventually make this more intuitive. At this time, further report and label system improvements are scheduled for next year, though we do not have an exact timeline planned at this time.

**Comment by FedorSteeman** [comment]
That is unfortunate to hear, because I am currently struggling to rewrite reports created in SpiReport for Specify6 and currently stuck with that. 

The trick I used by splitting and parsing aggregrated string contents is not entirely reliable and won't help me accessing other agent fields so I need something better than that. 

I have been experimenting with jasper report variables to pick out specific rows in the record set (like only those where loan agent role is something specific like "preparer". But because of the unusual way the report templates are set up, I cannot get this to work. 

Alternatively, I considered doing Specify7 API calls to get the loan's agent based on role, but not sure if that would work either.

**Comment by grantfitzsimmons** [comment]
>The trick I used by splitting and parsing aggregrated string contents is not entirely reliable and won't help me accessing other agent fields so I need something better than that.

I spent several hours trying to come up with a generic solution but I am facing the same issue as you. I'm going to discuss this issue further with the devs to come up with a real solution going forward as other users are similarly reliant on the scriptlet in Specify 6 to effectively build reports.

**Comment by FedorSteeman** [comment]
Would it be possible to pull in more data in the original query atomically so we can more directly access things like agents of different roles? As of now the aggregate fields are the only possibility, but it's not very reliable to analyse these during execution of the report script.

**Comment by tmalm** [comment]
Hi guys, 
Tobias from the Swedish Museum of Natural History here.
Not sure if you have solved these issues already?
We have also had to rely on Specify 6 for our loan reports but recently I managed to get a decent output of our loan forms from Specify 7 with the different agents involved, by rebuilding and restructuring the output in Jasper Studio (6.21.3). 

Instead of relying on the scriptlet I used the agent role as a determining group and rolled through them, while using sets of printWhenExpressions to get the correct agent information by role.
I think this could be a workaround for other scriptlet-depending reports as well, but depending on the layout needed, and since it is grouped by agent you wont be able to repeatedly print the different agents, and they have to be sorted in the correct way for the output.
But it solved our problem with loan invoices for now.

Attaching my jrxml code, the query code, and a couple of pdfs of the finished reports.
If this is interesting I can go into more detail how I built and structured it.

Cheers
Tobias

[EN Loan Rpt Q.json](https://github.com/user-attachments/files/22302136/EN.Loan.Rpt.Q.json)
[loan invoice 1st and 2nd borrower and institution.pdf](https://github.com/user-attachments/files/22302137/loan.invoice.1st.and.2nd.borrower.and.institution.pdf)
[loan invoice borrower and insitution.pdf](https://github.com/user-attachments/files/22302139/loan.invoice.borrower.and.insitution.pdf)
[Loan to Borrower.txt](https://github.com/user-attachments/files/22302170/Loan.to.Borrower.txt)

**Comment by tmalm** [comment]
Another way to deal with this from the jasper report side would be using subreports, connected to the main report.
Is this possible from the Specify 7 standpoint, and if so, where would those subreports be placed for them to be available from the main report when it is running?

**Comment by grantfitzsimmons** [comment]
@tmalm We currently do not have a mechanism to support subreports, but we would be interested to hear about any additional solutions you discover through experimentation!

**Reply by tmalm** [reply]
Thanks for the info about subreports, I couldnt find anything about it in the forum. Subreporting would be very useful for certain types of more complicated reports, but there are workarounds for some of them.
I will put that up as a future feature request if I cannot solve everything I need.

**Comment by grantfitzsimmons** [comment]
As an update for all interested in the `getByLoanAgentRole` scriptlet in particular:

In the [7.11.2 update](https://discourse.specifysoftware.org/t/specify-7-11-2-release-announcement/2800), we added 5 new agent fields to both the loan and gift tables to help mitigate the issue. You can use [Batch Edit](https://discourse.specifysoftware.org/t/batch-edit/2248) to move data from the `loanAgents` relationship to `agent1`, `agent2`, `agent3`, etc. so that the report or label query can be modified to use these fields instead. 

This eliminates the need to create any special text replacement expressions or `printWhenExpressions`.  At this time, we recommend updating and migrating to these fields over using the `loanAgents` relationship to make it possible to now return this information in a simplified and more structured way.

**Reply by tmalm** [reply]
Thanks for the reply, Grant!
That is awesome to hear, we are currently still on an older version, and I was not aware of these updates.
But it gave me the oppurtunity to learn Jasper Reports at least.
These changes should make the process much more flexible and straightforward.
I will look into it once we get the upgraded version.

**Comment by bronwyncombs** [comment]
sumPrepCounts is requested by PRI (Paleontological Research Institution)
